### DISCUSSION TITLE
Index and RQL for Full Text Search

### DISCUSSION THREAD

**Discussion by vijaywm** [ravendb/ravendb#12897] [General]
Hi

I have a document collection of Reviews, with a description field (text) and the ranking (int). The description field has a full text index on it.

```
class Review:
    description=""
    ranking = 0
```

Requirement: 
The output needed is description, rank and count of other reviews with a matching description.
This is the query, that gives me the results I need in Mysql, but does not scale up to millions of rows. 

```
select t.description, t.rank,
(
	select count(x.description) from `review` x 
	where x.description like concat('%',t.description,'%')
) match_count 
from `review` t
where t.rank < 1000
order by t.rank
```
I cannot figure out the index and RQL for this scenario. Is there a way to make this work for a collection with 50 million + records.

**Comment by ayende** [comment]
Using RQL, you can run:

`from Reviews where search(description, $term) and ranking < 1000 order by ranking desc`

And that would give you the results you want.

**Reply by vijaywm** [reply]
@ayende 

The output I need is e.g.
```
[
{"christmas tree", 300, 25},
{"cologne", 125, 2} ...
]
```
where "christmas tree" is the description, 300 is the rank and 25 is the number of docs which have description similar to "christmas tree".
So there is no parameter  $term, the only parameters are the min and max values for rank.. and the output is all docs within the rank along with the count of matching docs. 
I think your solution is filtering docs by the description, but I need all docs along with the count of other docs with similar description!

```
from Reviews as x
x.description, x.rank, **count_of_docs_matching_x_description**
where x.rank between 0 and 1000
order by x.rank
```

I have created an Index and RQL below is fine, except for the count of matching docs

```



// document

{
      "term": "waist trainer for women",
      "rank": 105,
      "start_date": "2021-08-15",
      "end_date": "2021-08-21",
      "@metadata": {
            "@collection": "Keyword"
      }
}

Index:

// map
from d in docs.keyword
select new
{
      start_date = d.start_date,
      end_date = d.end_date,
      term = d.term,
      rank = d.rank,
      occurance = 1,
      ranks = new object[0],
      average_rank = 0,
      keywords = new [] { new { d.Id, d.end_date }},
      keyword_id= ""
}

//reduce
from r in results
group r by new { r.term } into g
let total_occurance = g.Sum(x=>x.occurance)
select new {
      start_date = g.Min(x=>x.start_date),
      end_date = g.Max(x=>x.end_date),
      term = g.Key.term,
      rank = 0,
      occurance = total_occurance,
      ranks = g.Select(x=>x.rank),
      average_rank = g.Average(x=>x.rank)
      keywords= new object [0],
      keyword_id = g.SelectMany(x=>x.keywords).OrderByDescending(x=>x.end_date).Take(1).First().Id
}

// RQL

from index 'keyword/date/rank' as summary
where
summary.start_date between '2021-08-01' and '2021-08-31'
and
(
      search(summary.term ,'face masks', and )
      or search(summary.term ,'face mask', and )
)
and rank between 0 and 10000
load keyword_id as most_recent
select summary, most_recent
```

**Reply by ayende** [reply]
To get the count of matching values for this, you need to run a separate query.
You can use facets to run the same operation on the index, then query using:

```
from index 'keyword/date/rank'
where term in ($termsThatWerePreviouslyLoaded)
select facet(term)
```

That will give you the matches.
